File: src/main/java/cn/labzen/web/ap/internal/element/ElementField.java
package cn.labzen.web.ap.internal.element;

import cn.labzen.web.ap.internal.Utils;
import com.squareup.javapoet.TypeName;
import lombok.Getter;

import java.util.List;
import java.util.Objects;

@Getter
public final class ElementField implements Element {

  private final String name;
  private final TypeName type;
  private final List<ElementAnnotation> annotations;

  public ElementField(String name, TypeName type, List<ElementAnnotation> annotations) {
    this.name = name;
    this.type = type;
    this.annotations = annotations;
  }

  @Override
  public String keyword() {
    return Utils.getSimpleName(type) + " " + name;
  }

  @Override
  public String toString() {
    return keyword();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || getClass() != o.getClass()) return false;
    ElementField that = (ElementField) o;
    return Objects.equals(name, that.name) && Objects.equals(type, that.type);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, type);
  }
}

File: src/main/kotlin/cn/labzen/web/ap/internal/element/ElementField.kt
package cn.labzen.web.ap.internal.element

import cn.labzen.web.ap.internal.Utils
import com.squareup.javapoet.TypeName

data class ElementField(
  val name: String,
  val type: TypeName,
  val annotations: List<ElementAnnotation>,
) : Element {

  override fun keyword(): String =
    "${Utils.getSimpleName(type)} $name"

  override fun toString(): String =
    keyword()

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as ElementField

    if (name != other.name) return false
    if (type != other.type) return false

    return true
  }

  override fun hashCode(): Int {
    var result = name.hashCode()
    result = 31 * result + type.hashCode()
    return result
  }
}